### ISSUE TITLE
[Feature]: managing cnpg in GitOps workflows (FluxCD, ArgoCD ...)

### ISSUE THREAD

**Issue by relhajja** [cloudnative-pg/cloudnative-pg#5778]
### Is there an existing issue already for this feature request/idea?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new feature request to be evaluated.

### What problem is this feature going to solve? Why should it be added?

In kubernetes cluster manager using gitops tools like FluxCD, I faced a really annoying limiation, especially for restoring db cluster, which is the following:

When defining a CNPG cluster resource for a db, we specify the initialization mode for the cluster of the database. There are two options:

    initdb: This mode creates an empty database without any data.
    Recovery: This mode restores the database from a backup in object storage.

In the initdb mode, if the database crashes at some point and we need to restore it, we would have to restore it manually. However, since the database is managed by FluxCD, it will automatically recreate the cluster and initialize it without any data during the next reconciliation. This makes restoration impossible in this scenario it may seem to work momentarily, but all data will be lost shortly after (next FluxCD reconcialiation).

On the other hand, using the recovery mode is a good solution for restoring databases after a failure. However, for newly created databases, the CNPG cluster needs to be initialized in initdb mode. If we do this, we encounter the same problem as before—FluxCD will recreate an empty database, making it difficult to restore any existing data.

Another issue arises when the cluster recovers from a backup or is newly created. It must save new backups to a different path, as the old path contains backups from before the database crash. 

### Describe the solution you'd like

My goal is to manage the db and its restore in case of restore when changing the yaml of resource and go throught git and FluxCD.
The cnpg cluster must be more intellegent to create dynamic destination path for backup data, and also spec.bootstrap must have a schema that allow recovery from an empty backup and initialise db without data like initdb mode

### Describe alternatives you've considered

It would be more efficient if we added an option in the cluster resource to dynamically create the backup destination folder by concatenating the database name with the date of its creation, such as postgres-db-09-10-2020. This way, the new backup folder (e.g., postgres-db-09-10-2020) would start empty, allowing the database to be successfully initialized.

With this approach, everything is managed automatically, without needing to change the cluster resource code in the Git repository whenever we want to restore a database.

### Backport?

N/A

### Are you willing to actively contribute to this feature?

Yes

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by sxd** [comment]
Hello @riadelhajjaji-2001

This is not something to be solved at operator level, why you think the operator should solve that? every GitOps has it's own way to generated variables that can be used to replace, is not clear what wants to be solved here that cannot be solved using GitOps

Can you clarify a little bit?

Thanks

**Comment by onedr0p** [comment]
This would be awesome if supported, I know CPGO supports this...

Let me describe the scenario I run into which is closely related to @riadelhajjaji-2001 comment.

My Kubernetes cluster managed by [Flux](https://fluxcd.io/) is all happy, pg cluster running great, backups are running. Now there is a full kubernetes disaster and I need to re-provision my Kubernetes cluster. After I have the new Kubernetes cluster ready to go and before I tell Flux to put everything back, I need to change the CNPG Cluster manifest before it is applied by Flux as described in my comments in the manifest I posted above.

Ideally I would like CNPG to always try to recover on first bootstrap if there's a recovery found. It's the manually changing of the CNPG Cluster manifest that is hard to fit into a GitOps pattern. Currently I need to change the CNPG Cluster manifest, push it to git and then tell Flux to put everything back.

**Comment by PrivatePuffin** [comment]
@sxd This for sure is something that should be solved at the operator level.

Essentially CNPG is blocking reproducible builds for ALL gitops tools out there.

**Comment by m00nyONE** [comment]
+1 This is also our main problem with argo-cd at the moment. Even with preSync and postSync hooks, and a lot of helm shenanigans, it's not possible..

**Comment by urashidmalik** [comment]
Same here in my case i faced real dr on hetzner k8s, luckily i was running pgdumps which saved me.

This should be managed by operator it should have a state management to know that this was old cluster with existing data in objectstore and if data is missing on postgress side it performs restore automatically (configurable)

**Comment by PrivatePuffin** [comment]
@sxd Can we please actually get a response on the clarification you requested?

**Comment by ixxeL2097** [comment]
Facing the same challenge here, and support all commentaries in favor of GitOps. I really like CNPG but its lack of compatibility with GitOps methodologies has led me to seriously reconsider this approach. 
Would be awesome if CNPG could think its backup/restore features with GitOps compatibility in mind, enabling recovery operations to be fully declared in version-controlled manifests, rather than requiring manual interventions that break the GitOps paradigm

**Comment by PrivatePuffin** [comment]
@sxd @gbartolini  A response would be really appreciated.

**Comment by onedr0p** [comment]
Would this be (potentially) solved by setting the `cnpg.io/skipEmptyWalArchiveCheck` annotation on the cluster, or would that leave to some unforeseen issue?

**Comment by cormacrelf** [comment]
Can anyone write down a set of instructions for using CNPG through FluxCD safely? Or alternatively would you all recommend the cluster resource be managed outside of flux? Given what's written here I'll default to the latter, it seems too prone to accidents to put it in flux, in particular the part where flux recreates the cluster with initdb every 5 minutes sounds horrifying. But at least someone might save themselves a lot of pain if you folks can share your knowledge.

- Everyone mentions making manual changes to yaml during recovery. What changes do you need to make to `cluster.yaml` during recovery and when?
- Is the recovery process for full k8s failures different from PG-only cluster failures?
- Can some upfront work make things easier during recovery? For example
    1. create cluster with `bootstrap: initdb: ...`
    2. as soon as one backup is there, commit and push a change to `bootstrap: recovery: ...` with a configured object store. Flux reconciles this and CNPG either does nothing, or it recreates from the backup. (Anyone know what it will do there?)
    3. Send production traffic there, wait until you least expect it
    4. Observe DB failure
    5. Delete the cluster, 5 minutes later (or immediately if you call `flux reconcile`) flux reconciles it and bootstraps the new cluster from the named backup that you applied in step 2.

I kinda think my idea would have legs if it worked, because not only does restore happen without touching git, this workflow also means you have tested the restore process. But I don't know.

**Comment by bornav** [comment]
@cormacrelf for me it's 2 stages, when first initializing the database, and then converting it to use as recovery a backup solution

given this as a [reference](https://github.com/bornav/fluxcd/blob/main/kubernetes/apps/xauth/keycloak/new/pg-keycloak.yaml)

when first setting  up a database you would comment out the bootstrap.recovery and uncomment the init field

after the init is done you would uncomment it back to how it was

given this approach, when DR happens, it would recover the database from backup successfully without having to go trough init phase again


running this for ~1M and went trough multiple DR cycles, and had no data loss, the key is the annotation at the top

**Comment by cormacrelf** [comment]
Thank you @bornav that is great to know, I will run some tests to verify but this seems like a good workaround. This should probably go in the docs until there's something the operator can do.

**Comment by PrivatePuffin** [comment]
For @truecharts we're also making the `cnpg.io/skipEmptyWalArchiveCheck: "enabled"` tactic quoted above our default after initial bootstrapping, so like @bornav is saying indeed a 2 stage process, but at least it has seamless data recovery now.

**Comment by cormacrelf** [comment]
It seems the crucial thing to know about `cnpg.io/skipEmptyWalArchiveCheck` is that if you write backups for a new cluster to the same bucket location (which implies same serverName etc), then you'll be dealing with multiple recovery timelines. See the [Postgres Timelines docs][docs] for what this means. After postgres recovers, it generates a new timeline number and makes this part of the WAL archive filename, e.g. starting from 00000002, and you can control which timeline to restore from using the `targetTLI` field on `bootstrap.recovery`. You can also see what timeline a given cluster is up to by reading `kubectl get cluster mycluster -oyaml | yq .status.timelineID`. This timeline ID appears to increase monotonically every time you recover a cluster from a backup, no matter whether you are saving to a new bucket location or in the same spot.

I started writing a shell script to rotate the serverName for restored clusters using `flux suspend`, auto-edit YAML serverName, redeploy, `k delete cluster`, `flux resume`. But it turns out postgres is already very good at rotating the timeline ID with `cnpg.io/skipEmptyWalArchiveCheck` enabled, and it seems to be pretty much equivalent to rotating the serverName every time you restore. In fact if you do always rotate the serverName by incrementing an integer, you may as well call the revision number a "timelineID" as it will line up exactly, and all the postgres docs on timelines will apply to your rotated serverName.

The main thing to be careful of seems to be that you must not have two clusters backing up to the same location simultaneously, i.e. accidentally configuring two identical backup locations for entirely different clusters. The check does help prevent accidents like that.

[docs]: https://www.postgresql.org/docs/current/continuous-archiving.html#BACKUP-TIMELINES

**Comment by PrivatePuffin** [comment]
> I started writing a shell script to rotate the serverName for restored clusters using flux suspend, auto-edit YAML serverName, redeploy, k delete cluster, flux resume.

Yeah these things become a mess pretty quickly, thats why `cnpg.io/skipEmptyWalArchiveCheck` is a nicer solution.

> The main thing to be careful of seems to be that you must not have two clusters backing up to the same location simultaneously, i.e. accidentally configuring two identical backup locations for entirely different clusters. The check does help prevent accidents like that.

We've covered that issue in our charts by creating custom s3 paths based on chart, namespace, application and pvc naming.
So its practically made impossible for end-users to make that mistake :)

**Comment by tcurdt** [comment]
unstale

**Comment by spietras** [comment]
Both CNPG and GitOps are gaining popularity and many people will likely end up here (https://github.com/cloudnative-pg/cloudnative-pg/issues/5778, https://github.com/cloudnative-pg/cloudnative-pg/issues/4192, https://github.com/cloudnative-pg/cloudnative-pg/discussions/2250, https://github.com/cloudnative-pg/cloudnative-pg/discussions/6715) while researching all the options. So while this is still unresolved, let me ask all of you who were in a similar position and want to share your experience:

What did you end up doing? Did you find a satisfying workaround while still using CNPG? Or did you choose another Postgres operator that works nicely out of the box in this scenario?

**Comment by bornav** [comment]
@spietras in my case 
https://github.com/cloudnative-pg/cloudnative-pg/issues/5778#issuecomment-2782198691

**Comment by cormacrelf** [comment]
I think we've pretty much covered those questions in the discussion above. I would consider the issue resolved if someone wrote those findings up as documentation.

**Comment by bornav** [comment]
> I think we've pretty much covered those questions in the discussion above. I would consider the issue resolved if someone wrote those findings up as documentation.

yes and no, would still be nice to have a "if backup missing, run init" step so you are not forced to do a 2 step process of initializing

**Comment by spietras** [comment]
Ok, so seems that you can make it work with CNPG, but:
1. Cluster creation is a process with two steps
2. You need to have backups enabled
3. Cluster rebuild always does recovery instead of just using the data from a volume, like a normal reboot does

There must be some way to get rid of these limitations. I wonder how other Postgres operators deal with that...

**Comment by titilambert** [comment]
What about a mutating webhook (ie using kyverno) which patch the initdb job to:
1. Check if a pgdata folder exists
a. If doesn't exist, run initdb, as usual
b. If does exist, do nothing

**Comment by titilambert** [comment]
OK I just made kyverno mutation policy which works nicely !
With this policy, the init-db pod will check if the `$PGDATA` exists. If it exists, it does nothing, if not, it runs the original init-db command...

Here what you have to do:
1. Install kyverno: https://kyverno.io/docs/installation/methods/
2. Create a mutating policy (for each PG deployment you want this "feature") using the following template:
```
apiVersion: kyverno.io/v1
kind: Policy
metadata:
metadata:
  name: PGCLUSTER_NAME-mutation
  namespace: YOURNAMESPACE
spec:
  rules:
  - name: initdb-command
    match:
      any:
      - resources:
          kinds:
          - Job
          selector:
            matchLabels:
              cnpg.io/cluster: PGCLUSTER_NAME
              cnpg.io/jobRole: initdb
    mutate:
      patchStrategicMerge:
        spec:
          template:
            spec:
              containers:
              - (name): "*"
                env:
                - name: ORIGINAL_FULL_CMD
                  value: "{{ join('%%%', request.object.spec.template.spec.containers[0].command) }}"
                command:
                - /bin/bash
                - -c
                - |
                  set -e

                  echo "Checking if folder $PGDATA already exists"
                  if [ -d "$PGDATA" ]; then
                    echo "✓ Folder $PGDATA already exists, reusing it."
                    exit 0
                  fi

                  echo "✗ Folder $PGDATA doesn't exist yet"
                  echo "→ Run original CloudNativePG init..."
                  CLEANED_CMD=$(echo $ORIGINAL_FULL_CMD | sed  "s/%%%/ '/" | sed "s/%%%/' '/g")\'
                  echo "-> Original command: $CLEANED_CMD"
                  # Run original command
                  exec bash -c "$CLEANED_CMD"
```
3. Deploy your PG cluster

NOTE: If using fluxCD or ArgoCD, just integrate the Kyverno policy to your pg cluster deployment files...

WARNING: I'm using that on my personal server, NOT professional one. I did NOT test on multi node PG clusters. (I'm running only single cluster). 
If you can test it on a multi node cluster, please do it and write your feedback here. Thanks !

**Comment by samuelbartels20** [comment]
I am currently facing the same issue

**Comment by bornav** [comment]
@titilambert Thanks for sharing but my personal opinion is that having to add additional dependency to the cluster to get expected behavior, does not seem like a solution, but a workaround

**Comment by titilambert** [comment]
> [@titilambert](https://github.com/titilambert) Thanks for sharing but my personal opinion is that having to add additional dependency to the cluster to get expected behavior, does not seem like a solution, but a workaround

I do agree 100% !
The point is this work around is compatible with flux/... And fully automated .
That said, I still don't understand why the operator can not use the existing data in the pv and fallback to something if there is no data (or crash if the data are corrupted ...)

**Comment by NeuronButter** [comment]
unstale

**Comment by linwalth** [comment]
This is in the same vein as #5203 and I do think that this is an important feature as well for everyone deploying cnpg with gitops.

**Comment by michaelbeaumont** [comment]
Just realized I'm basically running into this with https://github.com/cloudnative-pg/cloudnative-pg/issues/9509
AFAICT @titilambert's solution could be integrated here https://github.com/cloudnative-pg/cloudnative-pg/blob/25b35a1725b0ad390ec88b085df51574f5f3ab3e/internal/cmd/manager/instance/initdb/cmd.go#L156

**Comment by bbaassssiiee** [comment]
https://clumpspirit.com/posts/cloudnativepg/

**Comment by bornav** [comment]
> https://clumpspirit.com/posts/cloudnativepg/

this is using soon to be deprecated intree barman instead of plugin, would not recommend this approach

**Comment by michaelbeaumont** [comment]
It also doesn't address the problem of how to initialize the DB if there isn't a backup yet.

**Comment by tcurdt** [comment]
Not using (client side) encryption for the backup is also an absolute no-go.

**Comment by bbaassssiiee** [comment]
Thanks for the heads up.

**Comment by shkarface** [comment]
I'm working on an admission mutating webhook which should make this fairly simple, I will share it here once I'm done

**Comment by lucian-student** [comment]
usntale

**Comment by linwalth** [comment]
Definitely still pertaining to our interests.

**Comment by lieberlois** [comment]
Totally agree with the "try recovery as default, if not existant, then init."

If somebody can point me to the relevant code location, I'd be happy to contribute :)

**Comment by linwalth** [comment]
I think it might be in here:

https://github.com/cloudnative-pg/cloudnative-pg/blob/9173c441c166aab39a55525b99933cf3cac75667/internal/controller/cluster_create.go#L1090

**Comment by lieberlois** [comment]
Thx @linwalth ! I guess we would just need some input from the maintainers if they agree with this general concept?

**Comment by linwalth** [comment]
@sxd do you think this has legs? :)

(Edit: Changed from gbartolini to sxd since this issue is assigned to sxd)

**Comment by linwalth** [comment]
Still not stale ;)

**Comment by 12345ieee** [comment]
I just got bit by this, I was terrified the data got lost.

I solved it by:
* stopping the operator
* killing the cluster pod
* going into the node, the pvc mount, and swapping the pgdata folders
* restarting the operator

This needed technically no gitOps level changes, but shortened my life by a few good years, can we get the "existing pgdata -> pgdata_numbers" behaviour under a flag? I'll personally contribute if need be.

**Comment by lucian-student** [comment]
@12345ieee 

That isnt great solution. I think this should be solved on operator level. 

Imo best solution would be add another strategy to InitDb.

InitDB would offer mixed strategy: where if there is backup available for example if u use barman plugin in s3 storage than it would load latest backup: otherwise it would fallback.

So maybe adding it as a fallback?

```
kind: Cluster
spec:
  bootstrap:
    recoveryInitDb:
        initDb: InitDbConfig
        recovery: RecoveryConfig
        lockFile: bucket/path/barman.lock
````

Barman before init would always check if ".lock" file is present in the bucket, if it is it would use recovery. Otherwise it would use initDb and create the .lock file in the S3 bucket.

This could be the solution. But it would probabably require change in CNPG and baraman cloud plugin.

**Comment by james-callahan** [comment]
Still an issue
